x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/ReplaceCaseWithInsensitiveMatch.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.esql.optimizer.rules.logical;
+
+import org.elasticsearch.common.lucene.BytesRefs;
+import org.elasticsearch.xpack.esql.core.expression.Expression;
+import org.elasticsearch.xpack.esql.core.expression.Literal;
+import org.elasticsearch.xpack.esql.core.expression.predicate.logical.Not;
+import org.elasticsearch.xpack.esql.core.expression.predicate.operator.comparison.BinaryComparison;
+import org.elasticsearch.xpack.esql.expression.function.scalar.string.ChangeCase;
+import org.elasticsearch.xpack.esql.expression.predicate.operator.comparison.Equals;
+import org.elasticsearch.xpack.esql.expression.predicate.operator.comparison.InsensitiveEquals;
+import org.elasticsearch.xpack.esql.expression.predicate.operator.comparison.NotEquals;
+
+public class ReplaceCaseWithInsensitiveMatch extends OptimizerRules.OptimizerExpressionRule<BinaryComparison> {
+
+    public ReplaceCaseWithInsensitiveMatch() {
+        super(OptimizerRules.TransformDirection.DOWN);
+    }
+
+    @Override
+    protected Expression rule(BinaryComparison bc) {
+        Expression e = bc;
+        if (bc.left() instanceof ChangeCase changeCase && bc.right().foldable()) {
+            if (bc instanceof Equals) {
+                e = replaceChangeCase(bc, changeCase);
+            } else if (bc instanceof NotEquals) { // not actually used currently, `!=` is built as `NOT(==)` already
+                e = new Not(bc.source(), replaceChangeCase(bc, changeCase));
+            }
+        }
+        return e;
+    }
+
+    private static Expression replaceChangeCase(BinaryComparison bc, ChangeCase changeCase) {
+        var foldedRight = BytesRefs.toString(bc.right().fold());
+        return changeCase.caseType().matchesCase(foldedRight)
+            ? new InsensitiveEquals(bc.source(), changeCase.field(), bc.right())
+            : Literal.FALSE;
```

Comment:
Use `Literal.of(bc, Boolean.FALSE)` to preserve the original source.
Additionally, see whether it makes sense to add a warning when dealing with a literal (WHERE TO_UPPER(field) == "lowercase").

Reply:
You could add an additional check to remove any nested casing function since only the top level one counts:
`to_upper(to_lower(to_upper(field)))` is the same as `to_upper(field)`.

Reply:
> You could add an additional check to remove any nested casing function

Added. 
It's basic; ideally, we'd be able to transform `TO_UPPER(CONCAT(TO_LOWER(field), "foo"))` to just `TO_UPPER(CONCAT(field, "foo"))`, but not `TO_UPPER((CONCAT(TO_LOWER(field), "-foo") == "x-foo")::KEYWORD)`. But maybe another PR.

Reply:
> see whether it makes sense to add a warning when dealing with a literal (WHERE TO_UPPER(field) == "lowercase")

We definitely can, but do we want then to expand a bit the scope and do this when other optimisations apply (folding to `null` or booleans), especially in filters?